src/games/strategy/engine/data/Route.java

```diff
@@ -355,8 +355,8 @@ public String toString() {
    *         the route consists of only a starting territory
    */
   public Territory getEnd() {
-    if (m_steps.size() == 0) {
-      return null;
+    if (m_steps.isEmpty()) {
+      return m_start;
```

Comment:
It has never made much sense for `Route` to keep track of the head of the route differently from the rest of it. From other perspectives we can see there are problems with that configuration. There are a lot of details that clients have to know about the implementation in order to set the class up properly and use it. 

I hit an error here pretty quickly in test data while adding unit tests. It does make more sense for us to return the end of a route so long as it has one territory in it. The choice to return null if there is no end is pretty arbitrary. 

There are a number of usages of this method. There are 4 things that are giving me confidence this is okay to change:
- all unit tests pass
- manual testing in-game looks good
- AI vs AI game (mix of hard + easy) played through round 12 of a game without any issues
- inspection of the usages for a solid hour didn't raise any flags

For inspection we can cut out a number of classes. Non-pro AI's we do not care so much about, so Pro AI usages were checked out. They all seemed okay. MoveValidator had a lot of usages, but for the most part we know that is going to be okay with this type of change, and much of that could be skimmed. The other locations all seemed to be okay as well.